File: plugins/plugin-memory/python/elizaos_plugin_memory/services/memory_service.py
"""Memory Service - Manages short-term and long-term memory."""

from __future__ import annotations

import logging
from datetime import datetime
from typing import Protocol, Optional
from uuid import UUID, uuid4

from elizaos_plugin_memory.types import (
    LongTermMemory,
    LongTermMemoryCategory,
    MemoryConfig,
    SessionSummary,
)

logger = logging.getLogger(__name__)


class DatabaseAdapter(Protocol):
    """Protocol for database operations."""

    async def insert(self, table: str, data: dict[str, object]) -> None:
        """Insert a record into the database."""
        ...

    async def select(
        self,
        table: str,
        conditions: dict[str, object],
        order_by: Optional[list[tuple[str, str]]] = None,
        limit: Optional[int] = None,
    ) -> list[dict[str, object]]:
        """Select records from the database."""
        ...

    async def update(
        self, table: str, data: dict[str, object], conditions: dict[str, object]
    ) -> None:
        """Update records in the database."""
        ...

    async def delete(self, table: str, conditions: dict[str, object]) -> None:
        """Delete records from the database."""
        ...


class CacheAdapter(Protocol):
    """Protocol for cache operations."""

    async def get(self, key: str) -> Optional[object]:
        """Get a value from cache."""
        ...

    async def set(self, key: str, value: object) -> None:
        """Set a value in cache."""
        ...


class MemoryService:
    """
    Memory Service.

    Manages both short-term (session summaries) and long-term (persistent facts) memory.
    """

    service_type: str = "memory"
    capability_description: str = (
        "Advanced memory management with short-term summarization and long-term persistent facts"
    )

    def __init__(
        self,
        config: Optional[MemoryConfig] = None,
        agent_id: Optional[UUID] = None,
        db: Optional[DatabaseAdapter] = None,
        cache: Optional[CacheAdapter] = None,
    ) -> None:
        """Initialize the memory service."""
        self.config = config or MemoryConfig()
        self.agent_id = agent_id
        self._db = db
        self._cache = cache
        self._session_message_counts: dict[UUID, int] = {}
        self._last_extraction_checkpoints: dict[str, int] = {}

    async def initialize(self, settings: dict[str, str]) -> None:
        """Initialize service from runtime settings."""
        if threshold := settings.get("MEMORY_SUMMARIZATION_THRESHOLD"):
            self.config.short_term_summarization_threshold = int(threshold)

        if retain := settings.get("MEMORY_RETAIN_RECENT"):
            self.config.short_term_retain_recent = int(retain)

        if interval := settings.get("MEMORY_SUMMARIZATION_INTERVAL"):
            self.config.short_term_summarization_interval = int(interval)

        if max_new := settings.get("MEMORY_MAX_NEW_MESSAGES"):
            self.config.summary_max_new_messages = int(max_new)

        long_term_enabled = settings.get("MEMORY_LONG_TERM_ENABLED")
        if long_term_enabled == "false":
            self.config.long_term_extraction_enabled = False
        elif long_term_enabled == "true":
            self.config.long_term_extraction_enabled = True

        if confidence := settings.get("MEMORY_CONFIDENCE_THRESHOLD"):
            self.config.long_term_confidence_threshold = float(confidence)

        if ext_threshold := settings.get("MEMORY_EXTRACTION_THRESHOLD"):
            self.config.long_term_extraction_threshold = int(ext_threshold)

        if ext_interval := settings.get("MEMORY_EXTRACTION_INTERVAL"):
            self.config.long_term_extraction_interval = int(ext_interval)

        logger.debug("MemoryService initialized with config: %s", self.config)

    async def stop(self) -> None:
        """Stop the service."""
        logger.info("MemoryService stopped")

    def get_config(self) -> MemoryConfig:
        """Get current configuration."""
        return MemoryConfig(
            short_term_summarization_threshold=self.config.short_term_summarization_threshold,
            short_term_retain_recent=self.config.short_term_retain_recent,
            short_term_summarization_interval=self.config.short_term_summarization_interval,
            long_term_extraction_enabled=self.config.long_term_extraction_enabled,
            long_term_vector_search_enabled=self.config.long_term_vector_search_enabled,
            long_term_confidence_threshold=self.config.long_term_confidence_threshold,
            long_term_extraction_threshold=self.config.long_term_extraction_threshold,
            long_term_extraction_interval=self.config.long_term_extraction_interval,
            summary_model_type=self.config.summary_model_type,
            summary_max_tokens=self.config.summary_max_tokens,
            summary_max_new_messages=self.config.summary_max_new_messages,
        )

    def update_config(self, **updates: object) -> None:
        """Update configuration."""
        for key, value in updates.items():
            if hasattr(self.config, key):
                setattr(self.config, key, value)

    def increment_message_count(self, room_id: UUID) -> int:
        """Increment and return message count for a room."""
        current = self._session_message_counts.get(room_id, 0)
        new_count = current + 1
        self._session_message_counts[room_id] = new_count
        return new_count

    def reset_message_count(self, room_id: UUID) -> None:
        """Reset message count for a room."""
        self._session_message_counts[room_id] = 0

    def _get_extraction_key(self, entity_id: UUID, room_id: UUID) -> str:
        """Generate cache key for extraction checkpoints."""
        return f"memory:extraction:{entity_id}:{room_id}"

    async def get_last_extraction_checkpoint(
        self, entity_id: UUID, room_id: UUID
    ) -> int:
        """Get the last extraction checkpoint for an entity in a room."""
        key = self._get_extraction_key(entity_id, room_id)

        if key in self._last_extraction_checkpoints:
            return self._last_extraction_checkpoints[key]

        if self._cache:
            try:
                checkpoint = await self._cache.get(key)
                message_count = int(checkpoint) if checkpoint else 0
                self._last_extraction_checkpoints[key] = message_count
                return message_count
            except Exception as e:
                logger.warning("Failed to get extraction checkpoint from cache: %s", e)

        return 0

    async def set_last_extraction_checkpoint(
        self, entity_id: UUID, room_id: UUID, message_count: int
    ) -> None:
        """Set the last extraction checkpoint for an entity in a room."""
        key = self._get_extraction_key(entity_id, room_id)
        self._last_extraction_checkpoints[key] = message_count

        if self._cache:
            try:
                await self._cache.set(key, message_count)
                logger.debug(
                    "Set extraction checkpoint for %s in room %s at count %d",
                    entity_id,
                    room_id,
                    message_count,
                )
            except Exception as e:
                logger.error("Failed to persist extraction checkpoint: %s", e)

    async def should_run_extraction(
        self, entity_id: UUID, room_id: UUID, current_message_count: int
    ) -> bool:
        """Check if long-term extraction should run."""
        threshold = self.config.long_term_extraction_threshold
        interval = self.config.long_term_extraction_interval

        if current_message_count < threshold:
            return False

        last_checkpoint = await self.get_last_extraction_checkpoint(entity_id, room_id)
        current_checkpoint = (current_message_count // interval) * interval
        should_run = current_message_count >= threshold and current_checkpoint > last_checkpoint

        logger.debug(
            "Extraction check: count=%d, threshold=%d, interval=%d, "
            "last_checkpoint=%d, current_checkpoint=%d, should_run=%s",
            current_message_count,
            threshold,
            interval,
            last_checkpoint,
            current_checkpoint,
            should_run,
        )

        return should_run

    async def store_long_term_memory(
        self,
        agent_id: UUID,
        entity_id: UUID,
        category: LongTermMemoryCategory,
        content: str,
        confidence: float = 1.0,
        source: Optional[str] = None,
        metadata: Optional[dict[str, object]] = None,
        embedding: Optional[list[float]] = None,
    ) -> LongTermMemory:
        """Store a long-term memory."""
        now = datetime.now()
        memory_id = uuid4()

        memory = LongTermMemory(
            id=memory_id,
            agent_id=agent_id,
            entity_id=entity_id,
            category=category,
            content=content,
            confidence=confidence,
            source=source,
            metadata=metadata or {},
            embedding=embedding,
            created_at=now,
            updated_at=now,
            access_count=0,
        )

        if self._db:
            await self._db.insert(
                "long_term_memories",
                {
                    "id": str(memory.id),
                    "agent_id": str(memory.agent_id),
                    "entity_id": str(memory.entity_id),
                    "category": memory.category.value,
                    "content": memory.content,
                    "metadata": memory.metadata,
                    "embedding": memory.embedding,
                    "confidence": memory.confidence,
                    "source": memory.source,
                    "access_count": memory.access_count,
                    "created_at": now,
                    "updated_at": now,
                },
            )

        logger.info("Stored long-term memory: %s for entity %s", category, entity_id)
        return memory

    async def get_long_term_memories(
        self,
        entity_id: UUID,
        category: Optional[LongTermMemoryCategory] = None,
        limit: int = 10,
    ) -> list[LongTermMemory]:
        """Retrieve long-term memories for an entity."""
        if not self._db or not self.agent_id:
            return []

        conditions: dict[str, object] = {
            "agent_id": str(self.agent_id),
            "entity_id": str(entity_id),
        }

        if category:
            conditions["category"] = category.value

        results = await self._db.select(
            "long_term_memories",
            conditions,
            order_by=[("confidence", "desc"), ("updated_at", "desc")],
            limit=limit,
        )

        return [
            LongTermMemory(
                id=UUID(str(row["id"])),
                agent_id=UUID(str(row["agent_id"])),
                entity_id=UUID(str(row["entity_id"])),
                category=LongTermMemoryCategory(str(row["category"])),
                content=str(row["content"]),
                metadata=dict(row.get("metadata", {})) if row.get("metadata") else {},
                embedding=list(row["embedding"]) if row.get("embedding") else None,
                confidence=float(row.get("confidence", 1.0)),
                source=str(row["source"]) if row.get("source") else None,
                created_at=row["created_at"] if isinstance(row["created_at"], datetime) else datetime.now(),
                updated_at=row["updated_at"] if isinstance(row["updated_at"], datetime) else datetime.now(),
                last_accessed_at=row.get("last_accessed_at") if isinstance(row.get("last_accessed_at"), datetime) else None,
                access_count=int(row.get("access_count", 0)),
            )
            for row in results
        ]

    async def update_long_term_memory(
        self,
        memory_id: UUID,
        entity_id: UUID,
        **updates: object,
    ) -> None:
        """Update a long-term memory."""
        if not self._db or not self.agent_id:
            return

        update_data: dict[str, object] = {"updated_at": datetime.now()}

        for key in ["content", "metadata", "confidence", "embedding", "last_accessed_at", "access_count"]:
            if key in updates:
                update_data[key] = updates[key]

        await self._db.update(
            "long_term_memories",
            update_data,
            {
                "id": str(memory_id),
                "agent_id": str(self.agent_id),
                "entity_id": str(entity_id),
            },
        )

        logger.info("Updated long-term memory: %s for entity %s", memory_id, entity_id)

    async def delete_long_term_memory(self, memory_id: UUID, entity_id: UUID) -> None:
        """Delete a long-term memory."""
        if not self._db or not self.agent_id:
            return

        await self._db.delete(
            "long_term_memories",
            {
                "id": str(memory_id),
                "agent_id": str(self.agent_id),
                "entity_id": str(entity_id),
            },
        )

        logger.info("Deleted long-term memory: %s for entity %s", memory_id, entity_id)

    async def get_current_session_summary(
        self, room_id: UUID
    ) -> Optional[SessionSummary]:
        """Get the current session summary for a room."""
        if not self._db or not self.agent_id:
            return None

        results = await self._db.select(
            "session_summaries",
            {
                "agent_id": str(self.agent_id),
                "room_id": str(room_id),
            },
            order_by=[("updated_at", "desc")],
            limit=1,
        )

        if not results:
            return None

        row = results[0]
        return SessionSummary(
            id=UUID(str(row["id"])),
            agent_id=UUID(str(row["agent_id"])),
            room_id=UUID(str(row["room_id"])),
            entity_id=UUID(str(row["entity_id"])) if row.get("entity_id") else None,
            summary=str(row["summary"]),
            message_count=int(row["message_count"]),
            last_message_offset=int(row.get("last_message_offset", 0)),
            start_time=row["start_time"] if isinstance(row["start_time"], datetime) else datetime.now(),
            end_time=row["end_time"] if isinstance(row["end_time"], datetime) else datetime.now(),
            topics=list(row.get("topics", [])) if row.get("topics") else [],
            metadata=dict(row.get("metadata", {})) if row.get("metadata") else {},
            embedding=list(row["embedding"]) if row.get("embedding") else None,
            created_at=row["created_at"] if isinstance(row["created_at"], datetime) else datetime.now(),
            updated_at=row["updated_at"] if isinstance(row["updated_at"], datetime) else datetime.now(),
        )

    async def store_session_summary(
        self,
        agent_id: UUID,
        room_id: UUID,
        summary: str,
        message_count: int,
        last_message_offset: int,
        start_time: datetime,
        end_time: datetime,
        entity_id: Optional[UUID] = None,
        topics: Optional[list[str]] = None,
        metadata: Optional[dict[str, object]] = None,
        embedding: Optional[list[float]] = None,
    ) -> SessionSummary:
        """Store a session summary."""
        now = datetime.now()
        summary_id = uuid4()

        session_summary = SessionSummary(
            id=summary_id,
            agent_id=agent_id,
            room_id=room_id,
            entity_id=entity_id,
            summary=summary,
            message_count=message_count,
            last_message_offset=last_message_offset,
            start_time=start_time,
            end_time=end_time,
            topics=topics or [],
            metadata=metadata or {},
            embedding=embedding,
            created_at=now,
            updated_at=now,
        )

        if self._db:
            await self._db.insert(
                "session_summaries",
                {
                    "id": str(session_summary.id),
                    "agent_id": str(session_summary.agent_id),
                    "room_id": str(session_summary.room_id),
                    "entity_id": str(session_summary.entity_id) if session_summary.entity_id else None,
                    "summary": session_summary.summary,
                    "message_count": session_summary.message_count,
                    "last_message_offset": session_summary.last_message_offset,
                    "start_time": session_summary.start_time,
                    "end_time": session_summary.end_time,
                    "topics": session_summary.topics,
                    "metadata": session_summary.metadata,
                    "embedding": session_summary.embedding,
                    "created_at": now,
                    "updated_at": now,
                },
            )

        logger.info("Stored session summary for room %s", room_id)
        return session_summary

    async def update_session_summary(
        self,
        summary_id: UUID,
        room_id: UUID,
        **updates: object,
    ) -> None:
        """Update a session summary."""
        if not self._db or not self.agent_id:
            return

        update_data: dict[str, object] = {"updated_at": datetime.now()}

        for key in ["summary", "message_count", "last_message_offset", "end_time", "topics", "metadata", "embedding"]:
            if key in updates:
                update_data[key] = updates[key]

        await self._db.update(
            "session_summaries",
            update_data,
            {
                "id": str(summary_id),
                "agent_id": str(self.agent_id),
                "room_id": str(room_id),
            },
        )

        logger.info("Updated session summary: %s for room %s", summary_id, room_id)

    async def get_session_summaries(
        self, room_id: UUID, limit: int = 5
    ) -> list[SessionSummary]:
        """Get session summaries for a room."""
        if not self._db or not self.agent_id:
            return []

        results = await self._db.select(
            "session_summaries",
            {
                "agent_id": str(self.agent_id),
                "room_id": str(room_id),
            },
            order_by=[("updated_at", "desc")],
            limit=limit,
        )

        return [
            SessionSummary(
                id=UUID(str(row["id"])),
                agent_id=UUID(str(row["agent_id"])),
                room_id=UUID(str(row["room_id"])),
                entity_id=UUID(str(row["entity_id"])) if row.get("entity_id") else None,
                summary=str(row["summary"]),
                message_count=int(row["message_count"]),
                last_message_offset=int(row.get("last_message_offset", 0)),
                start_time=row["start_time"] if isinstance(row["start_time"], datetime) else datetime.now(),
                end_time=row["end_time"] if isinstance(row["end_time"], datetime) else datetime.now(),
                topics=list(row.get("topics", [])) if row.get("topics") else [],
                metadata=dict(row.get("metadata", {})) if row.get("metadata") else {},
                embedding=list(row["embedding"]) if row.get("embedding") else None,
                created_at=row["created_at"] if isinstance(row["created_at"], datetime) else datetime.now(),
                updated_at=row["updated_at"] if isinstance(row["updated_at"], datetime) else datetime.now(),
            )
            for row in results
        ]

    async def get_formatted_long_term_memories(self, entity_id: UUID) -> str:
        """Get all long-term memories formatted for context."""
        memories = await self.get_long_term_memories(entity_id, None, 20)

        if not memories:
            return ""

        grouped: dict[LongTermMemoryCategory, list[LongTermMemory]] = {}
        for memory in memories:
            if memory.category not in grouped:
                grouped[memory.category] = []
            grouped[memory.category].append(memory)

        sections: list[str] = []
        for category, category_memories in grouped.items():
            category_name = category.value.replace("_", " ").title()
            items = "\n".join(f"- {m.content}" for m in category_memories)
            sections.append(f"**{category_name}**:\n{items}")

        return "\n\n".join(sections)

File: packages/typescript/src/advanced-memory/services/memory-service.ts
import { logger } from "../../logger.ts";
import {
	type IAgentRuntime,
	Service,
	type ServiceTypeName,
	type UUID,
} from "../../types/index.ts";
import type { MemoryStorageProvider } from "../../types/memory-storage.ts";
import type {
	LongTermMemory,
	LongTermMemoryCategory,
	MemoryConfig,
	SessionSummary,
} from "../types.ts";

export class MemoryService extends Service {
	static serviceType: ServiceTypeName = "memory" as ServiceTypeName;

	private sessionMessageCounts: Map<UUID, number>;
	private memoryConfig: MemoryConfig;
	private lastExtractionCheckpoints: Map<string, number>;

	/** Resolved at initialize(). null means no storage backend is available. */
	private storage: MemoryStorageProvider | null = null;

	capabilityDescription =
		"Memory management with short-term summarization and long-term persistent facts";

	constructor(runtime?: IAgentRuntime) {
		super(runtime);
		this.sessionMessageCounts = new Map();
		this.lastExtractionCheckpoints = new Map();
		this.memoryConfig = {
			shortTermSummarizationThreshold: 16,
			shortTermRetainRecent: 6,
			shortTermSummarizationInterval: 10,
			longTermExtractionEnabled: true,
			longTermVectorSearchEnabled: false,
			longTermConfidenceThreshold: 0.85,
			longTermExtractionThreshold: 30,
			longTermExtractionInterval: 10,
			summaryModelType: "TEXT_LARGE",
			summaryMaxTokens: 2500,
			summaryMaxNewMessages: 20,
		};
	}

	static async start(runtime: IAgentRuntime): Promise<Service> {
		const service = new MemoryService(runtime);
		await service.initialize(runtime);
		return service;
	}

	async stop(): Promise<void> {
		logger.info({ src: "service:memory" }, "MemoryService stopped");
	}

	async initialize(runtime: IAgentRuntime): Promise<void> {
		this.runtime = runtime;

		// Discover the storage provider registered by a database plugin.
		// If none exists, storage-backed features are disabled.
		const provider = runtime.getService(
			"memoryStorage",
		) as unknown as MemoryStorageProvider | null;
		if (!provider) {
			logger.warn(
				{ src: "service:memory", agentId: runtime.agentId },
				"No MemoryStorageProvider found — long-term memory and session summaries disabled. " +
					"Register a memoryStorage service from your database plugin to enable them.",
			);
		}
		this.storage = provider;

		// Read config overrides from environment / character settings.
		const threshold = runtime.getSetting("MEMORY_SUMMARIZATION_THRESHOLD");
		if (threshold) {
			this.memoryConfig.shortTermSummarizationThreshold = Number.parseInt(
				String(threshold),
				10,
			);
		}

		const retainRecent = runtime.getSetting("MEMORY_RETAIN_RECENT");
		if (retainRecent) {
			this.memoryConfig.shortTermRetainRecent = Number.parseInt(
				String(retainRecent),
				10,
			);
		}

		const summarizationInterval = runtime.getSetting(
			"MEMORY_SUMMARIZATION_INTERVAL",
		);
		if (summarizationInterval) {
			this.memoryConfig.shortTermSummarizationInterval = Number.parseInt(
				String(summarizationInterval),
				10,
			);
		}

		const maxNewMessages = runtime.getSetting("MEMORY_MAX_NEW_MESSAGES");
		if (maxNewMessages) {
			this.memoryConfig.summaryMaxNewMessages = Number.parseInt(
				String(maxNewMessages),
				10,
			);
		}

		const longTermEnabled = runtime.getSetting("MEMORY_LONG_TERM_ENABLED");
		if (longTermEnabled === "false" || longTermEnabled === false) {
			this.memoryConfig.longTermExtractionEnabled = false;
		} else if (longTermEnabled === "true" || longTermEnabled === true) {
			this.memoryConfig.longTermExtractionEnabled = true;
		}

		const confidenceThreshold = runtime.getSetting(
			"MEMORY_CONFIDENCE_THRESHOLD",
		);
		if (confidenceThreshold) {
			this.memoryConfig.longTermConfidenceThreshold = Number.parseFloat(
				String(confidenceThreshold),
			);
		}

		const extractionThreshold = runtime.getSetting(
			"MEMORY_EXTRACTION_THRESHOLD",
		);
		if (extractionThreshold) {
			this.memoryConfig.longTermExtractionThreshold = Number.parseInt(
				String(extractionThreshold),
				10,
			);
		}

		const extractionInterval = runtime.getSetting("MEMORY_EXTRACTION_INTERVAL");
		if (extractionInterval) {
			this.memoryConfig.longTermExtractionInterval = Number.parseInt(
				String(extractionInterval),
				10,
			);
		}

		logger.debug(
			{
				summarizationThreshold:
					this.memoryConfig.shortTermSummarizationThreshold,
				summarizationInterval: this.memoryConfig.shortTermSummarizationInterval,
				maxNewMessages: this.memoryConfig.summaryMaxNewMessages,
				retainRecent: this.memoryConfig.shortTermRetainRecent,
				longTermEnabled: this.memoryConfig.longTermExtractionEnabled,
				extractionThreshold: this.memoryConfig.longTermExtractionThreshold,
				extractionInterval: this.memoryConfig.longTermExtractionInterval,
				confidenceThreshold: this.memoryConfig.longTermConfidenceThreshold,
				storageAvailable: !!this.storage,
			},
			"MemoryService initialized",
			{ src: "service:memory" },
		);
	}

	// ── Helpers ──────────────────────────────────────────────────────────

	private requireStorage(): MemoryStorageProvider {
		if (!this.storage) {
			throw new Error(
				"MemoryStorageProvider not available. Register a memoryStorage service from your database plugin.",
			);
		}
		return this.storage;
	}

	getConfig(): MemoryConfig {
		return { ...this.memoryConfig };
	}

	updateConfig(updates: Partial<MemoryConfig>): void {
		this.memoryConfig = { ...this.memoryConfig, ...updates };
	}

	incrementMessageCount(roomId: UUID): number {
		const current = this.sessionMessageCounts.get(roomId) || 0;
		const newCount = current + 1;
		this.sessionMessageCounts.set(roomId, newCount);
		return newCount;
	}

	resetMessageCount(roomId: UUID): void {
		this.sessionMessageCounts.set(roomId, 0);
	}

	async shouldSummarize(roomId: UUID): Promise<boolean> {
		const count = await this.runtime.countMemories({
			roomIds: [roomId],
			unique: false,
			tableName: "messages",
		});
		return count >= this.memoryConfig.shortTermSummarizationThreshold;
	}

	private getExtractionKey(entityId: UUID, roomId: UUID): string {
		return `memory:extraction:${entityId}:${roomId}`;
	}

	async getLastExtractionCheckpoint(
		entityId: UUID,
		roomId: UUID,
	): Promise<number> {
		const key = this.getExtractionKey(entityId, roomId);

		const cached = this.lastExtractionCheckpoints.get(key);
		if (cached !== undefined) {
			return cached;
		}

		try {
			const checkpoint = await this.runtime.getCache<number>(key);
			const messageCount = checkpoint ?? 0;
			this.lastExtractionCheckpoints.set(key, messageCount);
			return messageCount;
		} catch (error) {
			const err = error instanceof Error ? error.message : String(error);
			logger.warn(
				{ src: "service:memory", err },
				"Failed to get extraction checkpoint from cache",
			);
			return 0;
		}
	}

	async setLastExtractionCheckpoint(
		entityId: UUID,
		roomId: UUID,
		messageCount: number,
	): Promise<void> {
		const key = this.getExtractionKey(entityId, roomId);
		this.lastExtractionCheckpoints.set(key, messageCount);

		try {
			await this.runtime.setCache(key, messageCount);
			logger.debug(
				{ src: "service:memory" },
				`Set extraction checkpoint for ${entityId} in room ${roomId} at count ${messageCount}`,
			);
		} catch (error) {
			const err = error instanceof Error ? error.message : String(error);
			logger.error(
				{ src: "service:memory", err },
				"Failed to persist extraction checkpoint to cache",
			);
		}
	}

	async shouldRunExtraction(
		entityId: UUID,
		roomId: UUID,
		currentMessageCount: number,
	): Promise<boolean> {
		const threshold = this.memoryConfig.longTermExtractionThreshold;
		const interval = this.memoryConfig.longTermExtractionInterval;

		if (currentMessageCount < threshold) {
			return false;
		}

		const lastCheckpoint = await this.getLastExtractionCheckpoint(
			entityId,
			roomId,
		);
		const currentCheckpoint =
			Math.floor(currentMessageCount / interval) * interval;
		const shouldRun =
			currentMessageCount >= threshold && currentCheckpoint > lastCheckpoint;

		logger.debug(
			{
				entityId,
				roomId,
				currentMessageCount,
				threshold,
				interval,
				lastCheckpoint,
				currentCheckpoint,
				shouldRun,
			},
			"Extraction check",
			{ src: "service:memory" },
		);

		return shouldRun;
	}

	// ── Storage operations (delegated to provider) ──────────────────────

	async storeLongTermMemory(
		memory: Omit<
			LongTermMemory,
			"id" | "createdAt" | "updatedAt" | "accessCount"
		>,
	): Promise<LongTermMemory> {
		const stored = await this.requireStorage().storeLongTermMemory(memory);
		logger.info(
			{ src: "service:memory" },
			`Stored long-term memory: ${stored.category} for entity ${stored.entityId}`,
		);
		return stored;
	}

	async getLongTermMemories(
		entityId: UUID,
		category?: LongTermMemoryCategory,
		limit = 10,
	): Promise<LongTermMemory[]> {
		if (limit <= 0) return [];
		return this.requireStorage().getLongTermMemories(
			this.runtime.agentId,
			entityId,
			{ category, limit },
		);
	}

	async updateLongTermMemory(
		id: UUID,
		entityId: UUID,
		updates: Partial<
			Omit<LongTermMemory, "id" | "agentId" | "entityId" | "createdAt">
		>,
	): Promise<void> {
		await this.requireStorage().updateLongTermMemory(
			id,
			this.runtime.agentId,
			entityId,
			updates,
		);
		logger.info(
			{ src: "service:memory" },
			`Updated long-term memory: ${id} for entity ${entityId}`,
		);
	}

	async deleteLongTermMemory(id: UUID, entityId: UUID): Promise<void> {
		await this.requireStorage().deleteLongTermMemory(
			id,
			this.runtime.agentId,
			entityId,
		);
		logger.info(
			{ src: "service:memory" },
			`Deleted long-term memory: ${id} for entity ${entityId}`,
		);
	}

	async getCurrentSessionSummary(roomId: UUID): Promise<SessionSummary | null> {
		return this.requireStorage().getCurrentSessionSummary(
			this.runtime.agentId,
			roomId,
		);
	}

	async storeSessionSummary(
		summary: Omit<SessionSummary, "id" | "createdAt" | "updatedAt">,
	): Promise<SessionSummary> {
		const stored = await this.requireStorage().storeSessionSummary(summary);
		logger.info(
			{ src: "service:memory" },
			`Stored session summary for room ${stored.roomId}`,
		);
		return stored;
	}

	async updateSessionSummary(
		id: UUID,
		roomId: UUID,
		updates: Partial<
			Omit<
				SessionSummary,
				"id" | "agentId" | "roomId" | "createdAt" | "updatedAt"
			>
		>,
	): Promise<void> {
		await this.requireStorage().updateSessionSummary(
			id,
			this.runtime.agentId,
			roomId,
			updates,
		);
		logger.info(
			{ src: "service:memory" },
			`Updated session summary: ${id} for room ${roomId}`,
		);
	}

	async getSessionSummaries(
		roomId: UUID,
		limit = 5,
	): Promise<SessionSummary[]> {
		return this.requireStorage().getSessionSummaries(
			this.runtime.agentId,
			roomId,
			limit,
		);
	}

	// ── Vector search (JS fallback; provider can override with native) ──

	async searchLongTermMemories(
		entityId: UUID,
		queryEmbedding: number[],
		limit = 5,
		matchThreshold = 0.7,
	): Promise<LongTermMemory[]> {
		if (limit <= 0) return [];
		if (!this.memoryConfig.longTermVectorSearchEnabled) {
			logger.warn(
				{ src: "service:memory" },
				"Vector search is not enabled, falling back to recent memories",
			);
			return this.getLongTermMemories(entityId, undefined, limit);
		}

		try {
			const candidates = await this.getLongTermMemories(
				entityId,
				undefined,
				200,
			);
			const scored: Array<{ memory: LongTermMemory; similarity: number }> = [];
			for (const memory of candidates) {
				if ((memory.embedding?.length ?? 0) === 0) continue;
				const similarity = cosineSimilarity(
					memory.embedding ?? [],
					queryEmbedding,
				);
				if (similarity < matchThreshold) continue;
				if (scored.length < limit) {
					scored.push({ memory, similarity });
					scored.sort((a, b) => b.similarity - a.similarity);
					continue;
				}
				if (similarity <= scored[scored.length - 1]?.similarity) continue;
				let index = 0;
				while (index < scored.length && scored[index].similarity > similarity) {
					index += 1;
				}
				scored.splice(index, 0, { memory, similarity });
				if (scored.length > limit) {
					scored.pop();
				}
			}
			return scored.map((x) => ({
				...x.memory,
				similarity: x.similarity,
			}));
		} catch (error) {
			logger.warn(
				{ error },
				"Vector search failed, falling back to recent memories",
				{ src: "service:memory" },
			);
			return this.getLongTermMemories(entityId, undefined, limit);
		}
	}

	// ── Formatting ──────────────────────────────────────────────────────

	async getFormattedLongTermMemories(entityId: UUID): Promise<string> {
		const memories = await this.getLongTermMemories(entityId, undefined, 20);
		if (memories.length === 0) return "";

		const grouped = new Map<LongTermMemoryCategory, LongTermMemory[]>();
		for (const memory of memories) {
			const existing = grouped.get(memory.category);
			if (existing) {
				existing.push(memory);
			} else {
				grouped.set(memory.category, [memory]);
			}
		}

		const sections: string[] = [];
		for (const [category, categoryMemories] of grouped.entries()) {
			const categoryName = category
				.split("_")
				.map((word) => word.charAt(0).toUpperCase() + word.slice(1))
				.join(" ");
			const items = categoryMemories.map((m) => `- ${m.content}`).join("\n");
			sections.push(`**${categoryName}**:\n${items}`);
		}

		return sections.join("\n\n");
	}
}

function cosineSimilarity(a: number[], b: number[]): number {
	if (a.length === 0 || b.length === 0 || a.length !== b.length) return 0;
	let dot = 0;
	let normA = 0;
	let normB = 0;
	for (let i = 0; i < a.length; i += 1) {
		const x = a[i] ?? 0;
		const y = b[i] ?? 0;
		dot += x * y;
		normA += x * x;
		normB += y * y;
	}
	const denom = Math.sqrt(normA) * Math.sqrt(normB);
	return denom === 0 ? 0 : dot / denom;
}
